### PR TITLE
15 implement verification

### PR DESCRIPTION
**Pull request by chadlagore**
**Status**: Ready for review.

# What Changed

+ Covered an additional test case to detect to transactions that already exist within the `BlockChain` (respend attack) - test case `TestTransactionRespend`.
+ Added some new helper methods: `BlockChain.AppendBlock` and `Block.ContainsTransaction`.
+ Refactored the tests.
> Refactor largely due to the fact that many of the tests were testing transactions that already existed within the `BlockChain`. While this is a reasonable exercise, we want to focus on testing _new_ transactions (transactions lacking consensus).

## How To Test Drive This PR

1. 👀 
2. 🏃 `go test`
3. 💵 profit

### CODE REVIEW

File: blockchain/block.go [hunk 1]
```diff
@@ -70,3 +70,14 @@ func DecodeBlock(r io.Reader) *Block {
 	gob.NewDecoder(r).Decode(&b)
 	return &b
 }
+
+// ContainsTransaction returns true and the transaction itself if the Block
+// contains the transaction.
+func (b *Block) ContainsTransaction(t *Transaction) bool {
```
Comment (jordanschalm) [inline blockchain/block.go]: 👍

File: blockchain/block.go [hunk 2]
```diff
@@ -70,3 +70,14 @@ func DecodeBlock(r io.Reader) *Block {
 	gob.NewDecoder(r).Decode(&b)
 	return &b
 }
+
+// ContainsTransaction returns true and the transaction itself if the Block
+// contains the transaction.
```
Comment (jordanschalm) [inline blockchain/block.go:75]: "and the transaction itself"?

File: blockchain/blockchain.go [hunk 3]
```diff
@@ -73,13 +73,20 @@ func (bc *BlockChain) ValidTransaction(t *Transaction) bool {
...
 
-	// Verify signature of T
+	// Verify signature of t
 	hash := HashSum(t.TxBody)
 	if !ecdsa.Verify(t.Sender.Key(), hash.Marshal(), t.Sig.R, t.Sig.S) {
 		return false
 	}
 
-	// Validate chain from input block to last block.
+	// Test if identical transaction already exits in chain.
```
Comment (jordanschalm) [inline blockchain/blockchain.go]: exists

File: blockchain/test_utils.go [hunk 4]
```diff
@@ -123,7 +123,8 @@ func newTransactionValue(s, r Wallet, a uint64) (*Transaction, error) {
 	return tbody.Sign(s, crand.Reader)
 }
 
-func newValidBlockChainFixture() *BlockChain {
+// newValidBlockChainFixture creates a valid blockchain of two blocks.
+func newValidBlockChainFixture() (*BlockChain, Wallet) {
```
Comment (jordanschalm) [inline blockchain/test_utils.go:127]: These test utils are awesome 🥇

**Review by jordanschalm — APPROVED** [review]
Couple of minor things. Awesome tests

File: blockchain/block.go (hunk 2 above)
Comment (chadlagore) [reply blockchain/block.go:75]: Seems like the right thing to do doesn't it. Breaks my `if block.ContainsTransaction(t) then` idiom 😢  Forgot the comment in there 😆

Comment (chadlagore) [reply blockchain/block.go:75]: 💡

**Pull request merged**